### ISSUE TITLE
Error when trying to access API via curl

### ISSUE THREAD

**Issue by everett-toews** [docker/docker#16107]
Description of problem:

When trying to access the API with curl, as documented in [Docker Remote API](https://docs.docker.com/reference/api/docker_remote_api/), I get an Unknown SSL protocol error in connection.

`docker version`:

```
$ docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:49:29 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:49:29 UTC 2015
 OS/Arch:      linux/amd64
```

`docker info`:

```
$ docker info
Containers: 0
Images: 0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: tmpfs
 Dirs: 0
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.9-boot2docker
Operating System: Boot2Docker 1.8.1 (TCL 6.3); master : 7f12e95 - Thu Aug 13 03:24:56 UTC 2015
CPUs: 1
Total Memory: 996.2 MiB
Name: test-curl
ID: TCSW:QYGQ:XDZJ:5XBL:MBA5:PCQL:F4RD:TFPW:HWMH:YMMM:SSOD:B2RF
Debug mode (server): true
File Descriptors: 9
Goroutines: 16
System Time: 2015-09-06T21:07:27.637987559Z
EventsListeners: 0
Init SHA1:
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Username: phymata
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

`uname -a`:

```
$ uname -a
Darwin MacBook-Pro.local 13.4.0 Darwin Kernel Version 13.4.0: Wed Mar 18 16:20:14 PDT 2015; root:xnu-2422.115.14~1/RELEASE_X86_64 x86_64
```

Environment details (AWS, VirtualBox, physical, etc.):

Mac 10.9.5
VirtualBox 5.0.0 r101573

How reproducible:

Very

Steps to Reproduce:
1. `docker-machine create --driver virtualbox test-curl`
2. `eval "$(docker-machine env test-curl)"`
3. `curl -v --insecure --cert $DOCKER_CERT_PATH/cert.pem --key $DOCKER_CERT_PATH/key.pem https://$(docker-machine ip test-curl):2376/images/json`

Actual Results:

```
* About to connect() to 192.168.99.102 port 2376 (#0)
*   Trying 192.168.99.102...
* Adding handle: conn: 0x7fc500803a00
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x7fc500803a00) send_pipe: 1, recv_pipe: 0
* Connected to 192.168.99.102 (192.168.99.102) port 2376 (#0)
* Unknown SSL protocol error in connection to 192.168.99.102:-9825
* Closing connection 0
curl: (35) Unknown SSL protocol error in connection to 192.168.99.102:-9825
```

Expected Results:

A JSON representation of Docker images.

**Comment by thaJeztah** [comment]
What version of curl are you using? I think I read somewhere that some older versions of curl had issues

**Comment by everett-toews** [comment]
Good tip.

```
$ curl --version
curl 7.30.0 (x86_64-apple-darwin13.0) libcurl/7.30.0 SecureTransport zlib/1.2.5
```

But if I install the latest according to brew and get really explicit and verbose about everything.

```
$ /usr/local/Cellar/curl/7.43.0/bin/curl -v --insecure --cert $DOCKER_CERT_PATH/cert.pem --key $DOCKER_CERT_PATH/key.pem https://$(docker-machine ip test-curl):2376/images/json

*   Trying 192.168.99.102...
* Connected to 192.168.99.102 (192.168.99.102) port 2376 (#0)
* WARNING: SSL: CURLOPT_SSLKEY is ignored by Secure Transport. The private key must be in the Keychain.
* WARNING: SSL: Certificate type not set, assuming PKCS#12 format.
* SSL: Can't load the certificate "/Users/everett/.docker/machine/machines/test-curl/cert.pem" and its private key: OSStatus -25299
* Closing connection 0
curl: (58) SSL: Can't load the certificate "/Users/everett/.docker/machine/machines/test-curl/cert.pem" and its private key: OSStatus -25299
```

Which is a bit odd to me because that cert definitely exists.

```
$ ls /Users/everett/.docker/machine/machines/test-curl/cert.pem
/Users/everett/.docker/machine/machines/test-curl/cert.pem
```

**Comment by thaJeztah** [comment]
Hm, this looks interesting;

    WARNING: SSL: CURLOPT_SSLKEY is ignored by Secure Transport. The private key must be in the Keychain.

But a quick Google search on the error-code; this looks to be the problem here; https://github.com/bagder/curl/issues/283

@nathanleclaire I know there were some issues around the certificates created by docker-machine; is this one being tracked already?

**Comment by everett-toews** [comment]
Ah ha. Poking around a bit I found this great post [curl the docker api on Mac (boot2docker)](http://www.craigjellick.com/curl-docker-api-mac-boot2docker/) by @cjellick 

Here's what I did to get it running

```
$ openssl pkcs12 -export -inkey $DOCKER_CERT_PATH/key.pem -in $DOCKER_CERT_PATH/cert.pem -name test-curl-client-side -out test-curl.p12 -password pass:mysecret

$ /usr/local/Cellar/curl/7.43.0/bin/curl -k --cert test-curl.p12:mysecret  https://$(docker-machine ip test-curl):2376/info
{"ID":"...","Containers":0,"Images":0, ...}
```

Note that if you try the above with stock Mac curl (v7.30.0 on my system) you still get the Unknown SSL protocol error in connection.

I'm hesitant to close this issue though. I think the [Docker Remote API](https://docs.docker.com/reference/api/docker_remote_api/) could be improved but I'm not sure the best way forward on that at the moment.
